File: src/main/scala/uk/gov/ons/methods/Melt.scala
package uk.gov.ons.methods

import org.apache.spark.sql.DataFrame

class Melt(val dfIn: DataFrame) extends BaseMethod {
  if (dfIn == null) throw new Exception("DataFrame cannot be null")

  import uk.gov.ons.methods.impl.MeltImpl._

  /** Scala function that calls version 1 of the melt method.
    *
    * @author stuart.russell@ext.ons.gov.uk
    * @param input      DataFrame        - The input DataFrame
    * @param id_vars    Seq[String]    - Column(s) which are used as unique identifiers
    * @param value_vars Seq[String] - Column(s) which are being unpivoted
    * @param var_name   String        - The name of a new column, which holds all
    *                   the value_vars names, defaulted to variable.
    * @param value_name String      - The name of a new column, which holds all the
    *                   values of value_vars column(s), defaulted to value.@param input
    * @return DataFrame
    */
  def melt1(input: DataFrame = dfIn, id_vars: Seq[String], value_vars: Seq[String],
            var_name: String = "variable", value_name: String = "value"): DataFrame = {
    mandatoryArgCheck(id_vars, value_vars)
    input.melt1(id_vars, value_vars, var_name, value_name)
  }

}

object Melt {
  def melt(df: DataFrame): Melt = new Melt(df)
}

File: python/sml/methods/melt.py
from pyspark.sql import DataFrame


class Melt:
    """
    This class contains methods which perform a melt (the reverse of a pivot) on a DataFrame.
    """

    def __init__(self, df=None):
        """
        Initialise function to instantiate the Melt class.

        :param df: a DataFrame
        """

        if df is None:
            raise TypeError

        self._df = df

        self._jMelt = self._df._sc._jvm.uk.gov.ons.api.java.methods.MeltFactory.melt(self._df._jdf)

    def melt1(self, df=None, id_vars=None, value_vars=None, var_name="variable", val_name="value"):
        """
        Version 1 of the melt function.

        :param df: a DataFrame
        :param id_vars: Column(s) which are used as unique identifiers
        :param value_vars: Column(s) which are being unpivoted
        :param var_name: The name of a new column, which holds all the value_vars names, defaulted to
                         variable.
        :param val_name: The name of a new column, which holds all the values of value_vars column(s),
                         defaulted to value.
        :return: pyspark.sql.DataFrame
        """

        if df is None:
            df = self._df

        return DataFrame(self._jMelt.melt1(df._jdf, id_vars, value_vars, var_name, val_name), df.sql_ctx)


def melt(df):
    return Melt(df)
